### PR TITLE
Using async-lock to prevent concurrent blockchain transactions

### PR DESCRIPTION
**Pull request by jrhender**
Need to have org creation be done in sequence.
This is because blockchain transactions for a single account are ordered and an account can only do one tx at a time.

- [x] add logs for locks acquisition
- [x] add logs for locks acquisition error
- [x] set lock acquisition timeout value as config
- [x] add new config values to documentation
- [x] test that lock is release if there is a failure of some kind

### CODE REVIEW

**Comment by jrhender** [comment]
@nichonien @artursudnik @JGiter have you every done in-memory locks before in nodejs (ideally nestjs)? Do you have any comments on this approach?
NestJs queues could be used here but I think that it's too heavy, I don't really want to add redis as a dependency.

**Review by JGiter — COMMENTED** [review]
What do you think of mutexing of sending transaction and signing of the iam-lib singer instead?

**Comment by jrhender** [comment]
> What do you think of mutexing of sending transaction and signing of the iam-lib singer instead?

Do you mean putting the lock/mutex in `IamService` instead of `OrgCreatorService`? If so, I think that's a good idea, thanks!

**Comment by JGiter** [comment]
> Do you mean putting the lock/mutex in `IamService` instead of `OrgCreatorService`?

Yes, this is seems more related to iam signer then org creator

File: src/modules/iam/iam.service.unit.spec.ts [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    });
+
+    it(`should execute concurrent calls to changeOrganizationOwnership in sequence`, async () => {
+      const blockchainOperations = ['1', '2'].map((newOwner) => {
+        return service.changeOrgOwnership({
+          namespace: '',
+          newOwner,
+        });
+      });
+      await Promise.all(blockchainOperations);
```
Comment (JGiter) [inline src/modules/iam/iam.service.unit.spec.ts:102]: It seems strange to me that test is passing. Both blockchain operations expect that `blockchainInUse` isn't set and then set it. So they should not be able to run concurrently. Do you know how both operations has managed to resolve?

**Review by JGiter — APPROVED** [review]
I see another opportunity for optimization by moving lock to `DomainReader`. This is also wouldn't require to write additional tests. We could only update existing one

**Comment by artursudnik** [comment]
I usually synchronized async code calls with the [async](https://caolan.github.io/async/v3/) library. I need to check how it relates to async-lock. If at all :)

**Comment by jrhender** [comment]
> I usually synchronized async code calls with the [async](https://caolan.github.io/async/v3/) library. I need to check how it relates to async-lock. If at all :)

Thanks for the note @artursudnik . `async` seems quite mature and highly used. It would be interesting if you think it can be used in this case

File: src/modules/iam/iam.service.unit.spec.ts (hunk 1 above)
Comment (jrhender) [reply src/modules/iam/iam.service.unit.spec.ts:102]: > So they should not be able to run concurrently. 

Yes, that's point of these tests and code change, that they **don't** run concurrently. Like they can be triggered/queued concurrently (a new request can be queued while waiting on the I/O to complete) but two I/O operations that involve the blockchain for the same account shouldn't be executed concurrently only one can succeed.

**Comment by jrhender** [comment]
> I see another opportunity for optimization by moving lock to DomainReader. This is also wouldn't require to write additional tests. We could only update existing one

@JGiter yes, I thought of putting the lock at the library level. Arguably, the spot where the lock code should be is `ethers` as that is where the transaction sending actually is. However, I think it is ultimately better that the lock is implemented at the application level because a lock at the library level would only work for single-machine application. If the application is spread across multiple machines, then shared lock/queue is needed (e.g. NestJs queues which use a Redis Instance). In that case, have a lock at the library level would be needless overhead and misleading as a process could acquire the lock but not actually have exclusive access to the resource. Is this how you would see it as well?

**Comment by JGiter** [comment]
> better that the lock is implemented at the application level because a lock at the library level would only work for single-machine application 

Thanks. I got it

File: src/modules/iam/iam.service.unit.spec.ts (hunk 1 above)
Comment (JGiter) [reply src/modules/iam/iam.service.unit.spec.ts:102]: Ah, thanks. I understood it now

**Pull request merged**